### PR TITLE
ENH: faster access to single elements in compressed sparse matrices

### PR DESCRIPTION
**Pull request by VarIr**
When compressed sparse matrices have sorted indices, binary search
can be used to access single elements. Access time complexity is
thus reduced from linear to logarithmic.
(Basically, I have created an implementation of the 
`# TODO make use of sorted indices (if present)`
comment in compressed.py)

This is my first PR to scipy, so please bear with me :)

### CODE REVIEW

File: scipy/sparse/compressed.py
Comment (perimosocordiae) [inline scipy/sparse/compressed.py]: This seems like more work than is needed. What about:
```python
minor_indices = self.indices[start:end]
insert_pos_left = np.searchsorted(minor_indices, minor_index, side='left')
insert_pos_right = np.searchsorted(minor_indices[insert_pos_left:], minor_index, side='right')
```

File: scipy/sparse/compressed.py
Comment (perimosocordiae) [inline scipy/sparse/compressed.py]: `self.dtype(0)` is cleaner here.

File: scipy/sparse/compressed.py
Comment (perimosocordiae) [inline scipy/sparse/compressed.py]: `minor_index` is a scalar, so I wouldn't worry about copies. Is this line necessary?

**Review by perimosocordiae — COMMENTED** [review]
Thanks for the PR! Just a few comments inline.

File: scipy/sparse/compressed.py
Comment (VarIr) [inline scipy/sparse/compressed.py]: During my initial benchmarks I observed ~linear time for accessing zero entries. Adding this line fixed that issue, so I assumed that copies were made. I haven't checked in detail, though.

File: scipy/sparse/compressed.py
Comment (VarIr) [inline scipy/sparse/compressed.py]: Good idea! I think we need to do the following, however:

`insert_pos_right = insert_pos_left + np.searchsorted(minor_indices[insert_pos_left:], minor_index, side='right')`

Otherwise, the indices do not match.

File: scipy/sparse/compressed.py
Comment (VarIr) [inline scipy/sparse/compressed.py]: Yes, that's more comprehensible. I'll change it to `self.dtype.type(0)`

**Comment by VarIr** [comment]
Thanks for the constructive review! I've incorporated your comments as well as simplified the second return statement a bit and added myself to THANKS.txt.

File: scipy/sparse/compressed.py
Comment (jaimefrio) [inline scipy/sparse/compressed.py]: You may have repeated indices and still have `self.has_sorted_indices` be `True`, e.g. the `4` in posiiton `(1, 2)` is the sum of two different entries in the `data` array:

```
a = sparse.csr_matrix((np.array([1.0, 2, 3, 2, 2, 5, 6]),
                       [1, 4, 5, 2, 2, 4, 3],
                       [0, 3, 6, 7]),
                      shape=(3, 6))
>>> a.has_sorted_indices
1
>>> a.A
array([[ 0.,  1.,  0.,  0.,  2.,  3.],
       [ 0.,  0.,  4.,  0.,  5.,  0.],
       [ 0.,  0.,  0.,  6.,  0.,  0.]])
```

So you need to add the values in between your left and right indices. Good thing is, this lets you join these two branches into a single one:

```
return self.data[start + insert_pos_left: start + insert_pos_right].sum(dtype=self.dtype)
```

File: scipy/sparse/compressed.py
Comment (jaimefrio) [inline scipy/sparse/compressed.py]: SciPy no longer supports versions of NumPy older than 1.7, but I think this comment is misleading... Unless @jnothman wants to chime in, he seems to have had in mind doing something like:

```
np.add.reduce(self.data[start:end], where=(minor_index == self.indices[start:end]), dtype=self.dtype)
```

Unfortunately, the `where=` kwarg is supported by ufuncs, but not by their methods, so the above doesn't work.

**Review by jaimefrio — CHANGES_REQUESTED** [review]
You current implementation does not cover all corner cases, see inline comments. And you most certainly need tests. But this is a cool enhancement, thanks for taking the time to work on it!

File: scipy/sparse/compressed.py
Comment (VarIr) [inline scipy/sparse/compressed.py]: Fair point, I haven't thought about such non-canonical forms. The return statement looks good to me.

**Comment by VarIr** [comment]
Thanks for your comments about corner cases! Could you please give me some guidance about the tests? I'm a little lost in `test_base.py`. There is `class _TestGetSet`, which seems to do appropriate tests. Given that my previous commit passed the tests, it seems that it does not test non-canonical forms (for which they should fail). How would I enable them?

File: scipy/sparse/compressed.py
Comment (jnothman) [inline scipy/sparse/compressed.py]: Sorry, not sure I can help here.

**Review by jnothman — COMMENTED** [review]
Hmm. The tests for non-canonical CSR/CSC should include `_TestGetSet`, I think.

File: scipy/sparse/compressed.py
Comment (jaimefrio) [inline scipy/sparse/compressed.py]: Blame says a three year younger version of yourself wrote that comment, so I thought it was worth trying...  I have trouble recognizing code I wrote last week, so I won't hold it against you. :P

**Comment by VarIr** [comment]
According to the output of `runtests.py -v`, they do. 
Looking at `_NonCanonicalMixin.spmatrix` makes me wonder, whether this method really inserts duplicate elements. I can't recognize any lines that would do that. Can anybody confirm?
(I have troubles setting up my debugging environment for scipy, so I'm stuck with looking at static code for now)

**Comment by jnothman** [comment]
_same_sum_duplicate inserts duplicate elements. But I guess it would be a
good idea to have an assertion to ensure they don't disappear somewhere
before testing, and thus through the per-format _arg1_for_noncanonical
methods.

**Comment by VarIr** [comment]
Previously, all tests for non-canonical compressed matrices had both duplicates and non-sorted indices. Due to the latter, the new code for binary search was never used in the tests, and thus they didn't fail.

I've added test classes for non-canonical matrices with sorted indices. They fail for my previous commit and succeed in the current one.

**Comment by perimosocordiae** [comment]
I agree that the commented out check is not necessary for this PR. Once that's out, I'm +1 to merge.

Did you try running the asv benchmarks to see the speedup? I'm not sure if the benchmark matrices have sorted indices, but if they do we should see the change.

**Comment by VarIr** [comment]
I can't get `runtests.py --bench-compare` to run. Anyway, considering my custom benchmarks, I assume the matrices in the scipy benchmarks are way too small for any visible improvements.

**Review by jaimefrio — APPROVED** [review]

**Comment by jaimefrio** [comment]
I don't fully follow the test scheme, but your implementation now LGTM. I will let someone else hit the merge button, though. Thanks!

**Comment by VarIr** [comment]
The idea was the following: `_TestGetSet.test_getelement`already provides a suitable test. It gets test data from `_NonCanonicalCompressedMixin` (w/ duplicate elements and non-sorted indices). The new class `_NonCanonicalSortedCompressedMixin` inherits from it, but overwrites `_arg1_for_noncanonical` so that the "un-sorting" does not take place. That's necessary, because the new code is in a `if self.has_sorted_indices` branch.

**Comment by jaimefrio** [comment]
I really don't want to be the one hitting the merge button without a dev more active in the project agreeing. But aside from someone verifying whether the tests are testing what we want them to, this should be ready to go.

**Comment by perimosocordiae** [comment]
My only remaining concern is that adding the sorted-indices non-canonical tests have exploded the sparse testcases more than we'd like. On master, `./runtests.py -s sparse` currently produces:
```
Ran 13263 tests in 188.572s

OK (KNOWNFAIL=44, SKIP=1803)
```

With the changes in this PR (on the same machine), I get:
```
Ran 15333 tests in 245.214s

OK (KNOWNFAIL=56, SKIP=1934)
```

That's a delta of +2070 tests just to cover 5 statements! The big problem, though, is that it's almost a minute slower to run the sparse tests.

**Comment by perimosocordiae** [comment]
This problem could be hacked around by overriding the `test_getelement` method in `_NonCanonicalCompressedMixin` to explicitly test both the sorted and non-sorted non-canonical cases, and removing the other changes to `test_base.py`.

This could be achieved by changing `_NonCanonicalCompressedMixin._arg1_for_noncanonical` to make index un-sorting conditional on the presence of a flag, then having `_NonCanonicalCompressedMixin.test_getelement` call the super `test_getelement` both with and without the flag set.

**Comment by VarIr** [comment]
I adapted the test scheme according to the suggestions above. However, instead of overriding `test_getelement` and `_arg1_for_noncanonical` in `_NonCanonicalCompressedMixin`, I did so in `_NonCanonicalCSMixin`. Otherwise, we'd run into problems with BSR matrices, which don't implement `getelement`.

File: scipy/sparse/tests/test_base.py
Comment (perimosocordiae) [inline scipy/sparse/tests/test_base.py]: I'm not sure why this condition exists. Seems like we could just call
```
self._arg1_for_noncanonical(M, sorted_indices=sorted_indices)
```
in either case.

File: scipy/sparse/tests/test_base.py
Comment (perimosocordiae) [inline scipy/sparse/tests/test_base.py]: I'd use a regular (non-keyword) argument here, because this is only getting called by the test runner anyway.

**Review by perimosocordiae — CHANGES_REQUESTED** [review]
I made some minor comments in-line, but overall the new approach to testing sorted non-canonicals is great.

It looks like `THANKS.txt` was updated on master, so you'll need to resolve the conflict. (Pro-tip: add your line somewhere other than the end of the list to reduce the odds of conflicting changes.)

File: scipy/sparse/tests/test_base.py
Comment (VarIr) [inline scipy/sparse/tests/test_base.py]: `TestCOONonCanonical` also inherits from `_NonCanonicalMixin`. Without this condition, its  `_arg1_for_noncanonical` gets too many arguments. We could do the following instead:
``` python
class TestCOONonCanonical(_NonCanonicalMixin, TestCOO):
    def _arg1_for_noncanonical(self, M, sorted_indices):
        ...
```
but I did not want to change the signature of methods that are not involved (`sorted_indices` would be unused here).
Another solution would be to override `spmatrix` in `_NonCanonicalCSMixin`, which would create a lot of code duplication.

File: scipy/sparse/tests/test_base.py
Comment (VarIr) [inline scipy/sparse/tests/test_base.py]: True, I'll change that.

File: scipy/sparse/tests/test_base.py
Comment (perimosocordiae) [inline scipy/sparse/tests/test_base.py]: I see. I think the cleanest solution is to make all subclasses accept the `sorted_indices` kwarg, even if they don't use it. Also I don't see a reason for the current duplication in `_arg1_for_noncanonical` between `_NonCanonicalCompressedMixin` and `_NonCanonicalCSMixin`. BSR should be tested with both sorted and unsorted indices, too.

Comment (VarIr) [inline scipy/sparse/tests/test_base.py]: Alright, will change that. I'm not sure about BSR, what do you think should be tested? It has no `get_element`. Testing everything with sorted and unsorted indices leads us back to my previous commit (too many tests).

**Review by perimosocordiae — APPROVED** [review]

**Comment by perimosocordiae** [comment]
Thanks for sticking with this PR! The Travis failure was unrelated to this change.

**Pull request merged**
